dohclient.go

```diff
@@ -275,66 +273,139 @@ func dohQuicTransport(endpoint string, opt DoHClientOptions) (http.RoundTripper,
+	quic.EarlyConnection
+
+	hostname  string
+	rAddr     string
+	lAddr     net.IP
+	tlsConfig *tls.Config
+	config    *quic.Config
+	mu        sync.Mutex
+	udpConn   *net.UDPConn
+}
+
+func newQuicConnection(hostname, rAddr string, lAddr net.IP, tlsConfig *tls.Config, config *quic.Config) (quic.EarlyConnection, error) {
+	connection, udpConn, err := quicDial(hostname, rAddr, lAddr, tlsConfig, config)
 	if err != nil {
 		return nil, err
 	}
-	pool.add(udpConn)
-	return quic.DialEarly(udpConn, udpAddr, hostname, tlsConfig, config)
+
+	Log.WithFields(logrus.Fields{
+		"protocol": "quic",
+		"hostname": hostname,
+		"remote":   rAddr,
+		"local":    lAddr.String(),
+	}).Debug("new quic connection")
+
+	return &quicConnection{
+		hostname:        hostname,
+		rAddr:           rAddr,
+		lAddr:           lAddr,
+		tlsConfig:       tlsConfig,
+		config:          config,
+		udpConn:         udpConn,
+		EarlyConnection: connection,
+	}, nil
 }
 
-// Wrapper for http3.RoundTripper due to https://github.com/quic-go/quic-go/issues/765
-// This wrapper will transparently re-open expired connections. Should be removed once the issue
-// has been fixed upstream.
-type http3ReliableRoundTripper struct {
-	*http3.RoundTripper
-	pool *udpConnPool
+func (s *quicConnection) OpenStreamSync(ctx context.Context) (quic.Stream, error) {
+	s.mu.Lock()
+	defer s.mu.Unlock()
+	stream, err := s.EarlyConnection.OpenStreamSync(ctx)
+	if netErr, ok := err.(net.Error); ok && (netErr.Timeout() || netErr.Temporary()) {
+		Log.WithError(err).Debug("temporary fail when trying to open stream, attempting new connection")
+		if err = quicRestart(s); err != nil {
+			return nil, err
+		}
+		stream, err = s.EarlyConnection.OpenStreamSync(ctx)
+	}
+	return stream, err
 }
 
-func (r *http3ReliableRoundTripper) RoundTrip(req *http.Request) (*http.Response, error) {
-	resp, err := r.RoundTripper.RoundTrip(req)
+func (s *quicConnection) OpenStream() (quic.Stream, error) {
+	s.mu.Lock()
+	defer s.mu.Unlock()
+	stream, err := s.EarlyConnection.OpenStream()
 	if netErr, ok := err.(net.Error); ok && (netErr.Timeout() || netErr.Temporary()) {
-		r.pool.closeAll()
-		r.RoundTripper.Close()
-		resp, err = r.RoundTripper.RoundTrip(req)
+		Log.WithError(err).Debug("temporary fail when trying to open stream, attempting new connection")
+		if err = quicRestart(s); err != nil {
+			return nil, err
+		}
+		stream, err = s.EarlyConnection.OpenStream()
 	}
-	return resp, err
+	return stream, err
 }
 
-// UDP connection pool. Also a workaround for for the http3.RoundTripper. When using a custom
-// dialer that open its own UDP connections, http3.RoundTripper doesn't close them when the
-// remote terminates a connection, or when calling Close(). So we have to keep track of the
-// connections and close them all before calling Close() on the http3.RoundTripper.
-type udpConnPool struct {
-	conns []*net.UDPConn
-	mu    sync.Mutex
+func (s *quicConnection) NextConnection() quic.Connection {
+	return nil
```

Comment:
Is this safe? I mean if it's called and used by the caller. Or should it return `s`?

Reply:
No, I wouldn't have thought so, the purpose of this function is to suggest a fallback connection downgrade (similar to downgrading from TLS v1.3 to TLS v1.2). Since we handle the one time fallback ourselves (with the quicReconnect) I thought it best to have no fallback (even falling back to ourselves, which would probably put us into a loop), hence returning `nil`. If I'm wrong then let me know and I'll send a clean up patch.

Reply:
You're probably right. That function isn't documented upstream, and luckily also not used from what I can see. There's a test that does this
```go
newConn := conn.NextConnection()
str, err := newConn.OpenUniStream()
```
It'd panic if `newConn` was nil, but again, it's not used so seems fine